controllers/class.tapestry-controller.php

```diff
@@ -0,0 +1,337 @@
+<?php
+require_once dirname(__FILE__) . "/../utilities/class.tapestry-errors.php";
+require_once dirname(__FILE__) . "/../utilities/class.tapestry-helpers.php";
+require_once dirname(__FILE__) . "/../utilities/class.tapestry-user-roles.php";
+require_once dirname(__FILE__) . "/../utilities/class.tapestry-node-permissions.php";
+require_once dirname(__FILE__) . "/../interfaces/interface.tapestry-controller.php";
+
+/**
+ * Add/update/retrieve a Tapestry
+ * 
+ */
+class TapestryController implements ITapestryController
+{
+    private $postId;
+
+    /**
+     * Constructor
+     * 
+     * @param   Number  $postId post ID
+     * 
+     * @return  NULL
+     */
+    public function __construct($postId = 0)
+    {
+        $this->postId = (int) $postId;
+    }
+
+    /**
+     * Add A Tapestry
+     * 
+     * @param   Object  $tapestry   Tapestry
+     * 
+     * @return  Object  $tapestry
+     */
+    public function save($tapestry)
+    {
+        if (empty($tapestry->settings)) {
+            return TapestryErrors::throwsError('SETTINGS_MISSING_IN_NEW_TAPESTRY');
+        }
+        if (!empty($tapestry->nodes)) {
+            return TapestryErrors::throwsError('NODES_EXIST_IN_NEW_TAPESTRY');
+        }
+        if (!empty($tapestry->groups)) {
+            return TapestryErrors::throwsError('GROUPS_EXIST_IN_NEW_TAPESTRY');
+        }
+        if (!empty($tapestry->links)) {
+            return TapestryErrors::throwsError('LINKS_EXIST_IN_NEW_TAPESTRY');
+        }
+
+        return $this->_addTapestry($tapestry);
+    }
+
+    /**
+     * Retrieve a Tapestry post
+     * 
+     * @return  Object  $tapestry
+     */
+    public function get()
+    {
+        if (!$this->postId) {
+            return TapestryErrors::throwsError('INVALID_POST_ID');
+        }
+
+        return $this->_getTapestry();
```

Comment:
The tapestry should automatically be loaded in the constructor. Here, you should not need to do any database operations. Rather, you should just be able to return the tapestry using the private properties of this class that were already set in the constructor.

Reply:
I don't think we can do this. It's because the `nodes` and `groups` are stored as arrays of IDs in the `tapestry`. When I load the data for the two properties, they will be arrays of IDs and not the actual data. Additionally, I would need to check various permissions to filter what to return to the client. This would require further further data operations in order to load the actual data after filtering.